### DISCUSSION TITLE
How to add corner radius for the PopupRoot and also using TransparencyLevelHint=AcrylicBlur?

### DISCUSSION THREAD

**Discussion by laolarou726** [AvaloniaUI/Avalonia#15471] [Q&A]
How to add a corner radius for the PopupRoot and also use TransparencyLevelHint=AcrylicBlur? Previously I used a Border to fake the  PopupRoot's corner radius. But it will still leave the right corners on the edges.

**Comment by maxkatz6** [comment]
I don't think it's possible. The only way would be to get access to the composition visual of WinUI elements tree directly.

**Reply by laolarou726** [reply]
I tried to apply the clip to the window but it doesn't work at all.

**Reply by maxkatz6** [reply]
Avalonia doesn't have an API for clipping native window. 
Anything clipping related in Avalonia is about drawn content inside of the window.

**Reply by laolarou726** [reply]
> Avalonia doesn't have an API for clipping native window. 
> 
> Anything clipping related in Avalonia is about drawn content inside of the window. 

I also tried to use the ExperimentalAcrylicBorder in the PopupRoot's template. But it is also not working at all? Is this by design or the limitations of the PopupRoot?

**Reply by maxkatz6** [reply]
I am not sure if ExperimentalAcrylicBorder is supposed to work with popups. I think it must be platform specific.

**Reply by maxkatz6** [reply]
In general, and properly, this task would be solved by giving user an access to the IContainerVisual handle (we can't expose COM API types, but IntPtr should be sufficient).

With that, user can build their own WinRT COM C# mapping (using CsWinRT or MicroCOM). Get strongly typed IContainerVisual object from the IntPtr we provide.
After which anything inside of the IContainerVisual can be clipped in whatever way you need. Including AcrylicBlur visual.

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/WinRT/Composition/WinUiCompositedWindow.cs#L67-L68

**Reply by maxkatz6** [reply]
Similarly to IMacOSTopLevelPlatformHandle, we can create:
```
    [Unstable]
    public interface IWinRTCompositedTopLevelHandle : IPlatformHandle
    {
        IntPtr Handle { get; } // the old HWnd
        IntPtr ContainerVisual { get; }
    }
```

**Reply by maxkatz6** [reply]
Alternatively, instead of ContainerVisual, ICompositionTarget pointer can be exposed.

**Reply by laolarou726** [reply]
> Alternatively, instead of ContainerVisual, ICompositionTarget pointer can be exposed. 

If PopupRoot can get the system's default corner radius that will be good enough for me.
